core/src/main/java/org/apache/calcite/rel/core/JoinRelType.java

```diff
@@ -152,4 +152,29 @@ public JoinRelType cancelNullsOnRight() {
   public boolean projectsRight() {
     return this != SEMI && this != ANTI;
   }
+
+  /** Returns whether this join type accepts pushing predicates from above into its predicate. */
+  public boolean canPushIntoFromAbove() {
+    return (this == INNER) || (this == SEMI);
+  }
+
+  /** Returns whether this join type accepts pushing predicates from above into its left input. */
+  public boolean canPushLeftFromAbove() {
+    return (this == INNER) || (this == LEFT) || (this == SEMI) || (this == ANTI);
+  }
+
+  /** Returns whether this join type accepts pushing predicates from above into its right input. */
+  public boolean canPushRightFromAbove() {
+    return (this == INNER) || (this == RIGHT);
+  }
+
+  /** Returns whether this join type accepts pushing predicates from within into its left input. */
+  public boolean canPushLeftFromWithin() {
+    return (this == INNER) || (this == RIGHT) || (this == SEMI);
+  }
+
+  /** Returns whether this join type accepts pushing predicates from within into its right input. */
+  public boolean canPushRightFromWithin() {
+    return (this == INNER) || (this == LEFT) || (this == SEMI);
+  }
```

Comment:
These new methods in `JoinRelType` are used only for feeding `RelOptUtil#classifyFilters` method so I am not sure if it is the best place to put them here but I don't have a better idea  at the moment. I would suggest to mark them as experimental to have some flexibility if we find a better approach in the near future. WDYT?

Reply:
An alternative could be adding these new methods as static methods in RelOptUtil receiving a Join or JoinRelType as parameter, perhaps?

Reply:
Added the experimental annotation.